File: Algorithms/T42_Trapping_Rain_Water.java
class Solution {
    public int trap(int[] height) {
        int n = height.length;
        if(n == 0) return 0;
        int i = 0, j = n - 1, leftMax = height[0], rightMax = height[n - 1], res = 0;
        while(i <= j){
            if(height[i] <= height[j]){
                if(height[i] >= leftMax) leftMax = height[i];
                else res += leftMax - height[i];
                i++;
            }else{
                if(height[j] >= rightMax) rightMax = height[j];
                else res += rightMax - height[j];
                j--;
            }
        }
        return res;
    }
}

File: Algorithms/C++/T42_Trapping_Rain_Water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        stack<int> s;
        int res = 0;
        for(int i = 0; i < height.size(); i++){
            int h = height[i];
            while(!s.empty() && h > height[s.top()]){
                int idx = s.top();
                s.pop();
                if(s.empty()) break;
                res += (min(h, height[s.top()]) - height[idx]) * (i - s.top() - 1);
            }
            s.push(i);
        }
        return res;
    }
};
